lib/AST/ExtInfo.cpp

```diff
@@ -64,14 +54,99 @@ void ClangTypeInfo::dump(llvm::raw_ostream &os,
   }
 }
 
+// MARK: - UnexpectedClangTypeError
+
+Optional<UnexpectedClangTypeError> UnexpectedClangTypeError::checkClangType(
+  SILFunctionTypeRepresentation silRep,
+  const clang::Type *type, bool expectNonnullForCOrBlock, bool expectCanonical) {
+#ifdef NDEBUG
+  return None;
+#else
+  bool isBlock = true;
+  switch (silRep) {
+  case SILFunctionTypeRepresentation::CFunctionPointer:
+      isBlock = false;
+      LLVM_FALLTHROUGH;
+  case SILFunctionTypeRepresentation::Block: {
+    if (!type) {
+      if (expectNonnullForCOrBlock)
+        return {{Kind::NullForCOrBlock, type}};
+      return None;
+    }
+    if (expectCanonical && !type->isCanonicalUnqualified())
+      return {{Kind::NonCanonical, type}};
+    if (isBlock && !type->isBlockPointerType())
+      return {{Kind::NotBlockPointer, type}};
+    if (!isBlock && !(type->isFunctionPointerType()
+                      || type->isFunctionReferenceType()))
+      return {{Kind::NotFunctionPointerOrReference, type}};
+    return None;
+  }
+  default: {
+    if (type)
+      return {{Kind::NonnullForNonCOrBlock, type}};
+    return None;
+  }
+  }
+#endif
+}
+
+void UnexpectedClangTypeError::dump() {
+  auto &e = llvm::errs();
+  using Kind = UnexpectedClangTypeError::Kind;
+  switch (errorKind) {
+  case Kind::NullForCOrBlock: {
+    e << "Expected non-null Clang type for @convention(c)/@convention(block)"
+      << " function but found nullptr.";
+    break;
+  }
+  case Kind::NonnullForNonCOrBlock: {
+    e << ("Expected null Clang type for non-@convention(c),"
+          " non-@convention(block) function but found:\n");
+    type->dump();
+    break;
+  }
+  case Kind::NotBlockPointer: {
+    e << ("Expected block pointer type for @convention(block) function but"
+          " found:\n");
+    type->dump();
+    break;
+  }
+  case Kind::NotFunctionPointerOrReference: {
+    e << ("Expected function pointer/reference type for @convention(c) function"
+          " but found:\n");
+    type->dump();
+    break;
+  }
+  case Kind::NonCanonical: {
+    e << "Expected canonicalized Clang type but found:\n";
+    type->dump();
+    break;
+  }
+  llvm_unreachable("Unhandled case for UnexpectedClangTypeError");
```

Comment:
This isn't the right place for an unreachable; it needs to go after the switch, to "cover" the implicit exit when a matching case isn't present in the switch.  Here it isn't doing anything.  This means that the cases need to `return` instead of `break`.